In [18]:
!pip install -r requirements.txt

In [2]:
import calitp
from calitp.tables import tbl
from siuba import *

import pandas as pd
import numpy as np
import geopandas as gpd
import fiona

from ipyleaflet import Map, GeoJSON, projections, basemaps, GeoData, LayersControl, WidgetControl, GeoJSON
from ipywidgets import Text, HTML

import shapely
from shapely.geometry import LineString, MultiPoint
from shapely.ops import split, substring

import zlib
import datetime as dt

from utilities import *

/opt/conda/lib/python3.9/site-packages/geopandas/_compat.py:111: UserWarning: The Shapely GEOS version (3.9.1-CAPI-1.14.2) is incompatible with the GEOS version PyGEOS was compiled with (3.10.1-CAPI-1.16.0). Conversions between both will be slow.
  warnings.warn(


### High Quality Transit Areas Relevant Statutes

[PRC 21155](https://leginfo.legislature.ca.gov/faces/codes_displaySection.xhtml?sectionNum=21155.&lawCode=PRC)
* Major transit stop definition: _A major transit stop is as defined in Section 21064.3, except that, for purposes of this section, it also includes major transit stops that are included in the applicable regional transportation plan_
* High-quality transit corridor definition: _For purposes of this section, a high-quality transit corridor means a corridor with fixed route bus service with service intervals no longer than 15 minutes during peak commute hours._
    * Unable to locate definition of "peak commute hours"

[PRC 21064.3](https://leginfo.legislature.ca.gov/faces/codes_displaySection.xhtml?sectionNum=21064.3.&lawCode=PRC)
* _Major transit stop means a site containing any of the following:
(a) An existing rail or bus rapid transit station.
(b) A ferry terminal served by either a bus or rail transit service.
(c) The intersection of two or more major bus routes with a frequency of service interval of 15 minutes or less during the morning and afternoon peak commute periods._
    * "Intersection" may not be sufficiently well-defined for this analysis

[PRC 21060.2](https://leginfo.legislature.ca.gov/faces/codes_displaySection.xhtml?lawCode=PRC&sectionNum=21060.2.&highlight=true&keyword=bus%20rapid%20transit)
* _(a) “Bus rapid transit” means a public mass transit service provided by a public agency or by a public-private partnership that includes all of the following features:
(1) Full-time dedicated bus lanes or operation in a separate right-of-way dedicated for public transportation with a frequency of service interval of 15 minutes or less during the morning and afternoon peak commute periods.
(2) Transit signal priority.
(3) All-door boarding.
(4) Fare collection system that promotes efficiency.
(5) Defined stations._
    * Unlikely to determine if a service qualifies as BRT under this definition using GTFS alone

In [3]:
# ##definition of "peak" pending
# am_peak = range(6, 9)
# pm_peak = range(16, 19)

In [4]:
# am_commute_hours = list(am_peak)
# pm_commute_hours = list(pm_peak)
# commute_hours = am_commute_hours + pm_commute_hours

In [5]:
def create_segments(geometry):
    '''Splits a Shapely LineString into smaller LineStrings. If a MultiLineString passed,
    splits each LineString in that collection. 
    '''
    
    lines = []
    segment_distance_meters = 1250
    geometry = geometry.iloc[0]
    if hasattr(geometry, 'geoms'): ##check if MultiLineString
        linestrings = geometry.geoms
    else:
        linestrings = [geometry]
    for linestring in linestrings:
        for i in range(0, int(linestring.length), segment_distance_meters):
            lines.append(substring(linestring, i, i+segment_distance_meters))
    return lines

In [6]:
def find_stop_with_high_trip_count(segment, stops, stop_times, rank, calculated_stops):
    '''Given a shape segment, finds the stop serving the most (or other rank) trips within that segment.
    Adds that stop's stop_id to segment data (a row).
    '''
    
    stops_in_seg = gpd.clip(stops, segment.geometry)
    if stops_in_seg.size == 0:
        return segment
    
    stop_times_in_seg = stops_in_seg >> inner_join(_, stop_times, on = 'stop_id')
    trip_count_by_stop = stop_times_in_seg >> count(_.stop_id) >> arrange(-_.n) >> rename(n_trips = _.n)
    try:
        stop_id = trip_count_by_stop['stop_id'].iloc[rank - 1]

        if stop_id in list(calculated_stops):
            return segment
        segment['stop_id'] = stop_id
        segment['n_trips'] = trip_count_by_stop['n_trips'].iloc[rank - 1]
        return segment
    except IndexError:
        return segment

In [7]:
def get_operator_views(itp_id):
    '''Returns relevant views from the data warehouse for a single transit operator.
    '''
    shapes = tbl.gtfs_schedule.shapes() >> filter(_.calitp_itp_id == int(itp_id)) >> collect()
    shapes = gpd.GeoDataFrame(shapes, 
                              geometry = gpd.points_from_xy(shapes.shape_pt_lon, shapes.shape_pt_lat),
                              crs = 'EPSG:4326').to_crs('EPSG:6414') ## https://epsg.io/6414 (meters)
    print('loaded shapes')
    if itp_id != 273:
        wednesday = (tbl.views.gtfs_schedule_fact_daily_service() 
                 >> filter(_.calitp_itp_id == int(itp_id)) >> collect())
        wednesday = wednesday >> arrange(-_.service_date)
        wednesday = wednesday[wednesday['service_date'].apply(lambda x: x.weekday() == 2)]
        wednesday = wednesday[wednesday['service_date'].apply(lambda x: x.month == 12)] ## December
        wednesday = wednesday[wednesday['service_date'] == wednesday['service_date'].iloc[0]] ## pick most recent Wednesday from Dec or earlier
        wednesday = wednesday >> select(_.calitp_itp_id, _.calitp_url_number, _.service_id)
    
    bus_routes = (tbl.gtfs_schedule.routes()
                    >> filter(_.calitp_itp_id == int(itp_id))
                    >> filter((_.route_type == '3') | (_.route_type == '11'))
                    >> select(_.calitp_itp_id, _.calitp_url_number, _.route_id) 
                    # >> collect()
                 )
    print('loaded bus routes')
    
    if itp_id == 273:
        trips = (tbl.gtfs_schedule.trips()
         >> filter(_.calitp_itp_id == int(itp_id))
         >> filter(_.service_id == '1') ## temporary hardcode for SacRT
         >> inner_join(_, bus_routes, on = ['calitp_itp_id', 'calitp_url_number', 'route_id'])
         >> collect())
    else:
        trips = (tbl.gtfs_schedule.trips()
                 >> filter(_.calitp_itp_id == int(itp_id))
                 >> inner_join(_, bus_routes, on = ['calitp_itp_id', 'calitp_url_number', 'route_id'])
                 >> collect()
                 >> inner_join(_, wednesday, on = ['calitp_itp_id', 'calitp_url_number', 'service_id'])
                 )
    print('loaded trips')
    stop_times = (tbl.gtfs_schedule.stop_times()
                  >> filter(_.calitp_itp_id == int(itp_id))
                  >> collect())
    stop_times =  (stop_times >> inner_join(_, trips, on = ['calitp_itp_id', 'calitp_url_number', 'trip_id'])
                  >> select(-_.stop_headsign, -_.pickup_type, -_.drop_off_type,
                            -_.continuous_pickup, -_.continuous_drop_off, -_.shape_dist_travelled,
                           -_.timepoint)
                  )
    print('loaded stop times')

    stops = (tbl.gtfs_schedule.stops() 
             >> filter(_.calitp_itp_id == itp_id)
             >> select(_.stop_id, _.stop_lat, _.stop_lon)
             >> collect())
    stops = gpd.GeoDataFrame(stops,
                     geometry = gpd.points_from_xy(stops.stop_lon, stops.stop_lat),
                     crs = 'EPSG:4326').to_crs('EPSG:6414') ## https://epsg.io/6414 (meters)
    print('loaded stops')

    return shapes, trips, stop_times, stops

In [13]:
def fix_arrival_time(gtfs_timestring):
    '''Reformats a GTFS timestamp (which allows the hour to exceed 24 to mark service day continuity)
    to standard 24-hour time.
    '''
    split = gtfs_timestring.split(':')
    hour = int(split[0])
    if hour >= 24:
        split[0] = str(hour - 24)
        corrected = (':').join(split)
        return corrected.strip()
    else:
        return gtfs_timestring.strip()

In [14]:
def single_shape_hqta(shapes, trips, stop_times, stops, route_count_by_stop, shape_id, hqta):
    '''Starting with a single GTFS shape, split that shape into segments and determine if each segment qualifies
    as an HQTA. Existing segments within a shape are dropped for that shape, since their peak frequency and 
    HQTA status would have already been calculated for a previous shape.
    '''
    
    single_shape = (shapes
         >> filter(_.shape_id == shape_id)
         >> mutate(shape_pt_sequence = _.shape_pt_sequence.astype('int64'))
         >> arrange(_.shape_pt_sequence) ##arrange, then convert to line to preserve order...
        )
    
    route_line = LineString(list(single_shape['geometry']))
    single_line = single_shape[['calitp_itp_id', 'shape_id', 'calitp_extracted_at']].iloc[[0]] ##preserve info cols
    single_line['geometry'] = route_line
    single_line = gpd.GeoDataFrame(single_line, crs='EPSG:6414')
    
    if shape_id in debug_ids:
        print(f'***debug shape*** {shape_id}')
        debug_dict[f'{shape_id}_single_line'] = single_line
        debug_dict[f'{shape_id}_hqta'] = hqta

    calculated_stops = []
    if hqta.size != 0:
        already_calculated = hqta.dissolve(by='calitp_itp_id') ## get single polygon of HQTA calculation complete area
        single_line = single_line.overlay(already_calculated, how='difference') ## drop calculation complete area from current shape
        if single_line.size == 0:
            segments_with_max_stop = None
            print(f'no line for shape {shape_id}')
            return

    segmented = pd.DataFrame()
    for segment in create_segments(single_line.geometry):
        to_append = single_line.drop(columns=['geometry'])
        to_append['geometry'] = segment
        segmented = segmented.append(to_append)
        
    segmented = segmented.reset_index()
    segmented['segment_sequence'] = segmented.index.astype(str)
    segmented = segmented.astype({'calitp_itp_id': str})
    
    ## compute (hopefully unique) hash of segment id that can be used across routes/operators
    segmented['hqta_segment_id'] = segmented.apply(lambda x:
                zlib.crc32((x.calitp_itp_id + x.shape_id + x.segment_sequence).encode('utf-8')), axis=1)
    
    segmented.geometry = segmented.buffer(50) ##generous buffer for street/sidewalk width? Required to spatially find stops within each segment

    if shape_id in debug_ids:
        debug_dict[f'{shape_id}_segmented'] = segmented
        debug_dict[f'{shape_id}_stops'] = stops
        debug_dict[f'{shape_id}_route_ct_by_stop'] = route_count_by_stop
     
    segments_with_max_stop = segmented.apply(find_stop_with_high_trip_count, axis=1,
                                             args = (stops, stop_times, 1, calculated_stops))
    
    if not 'stop_id' in segments_with_max_stop.columns:
        segments_with_top2_stops = None
        print(f'no stops for shape {shape_id}')
        return ## no stops within segment

    max_stop_times = (stop_times 
                 >> select(_.stop_id, _.trip_id, _.departure_time)
                 >> inner_join(_, segments_with_max_stop, on = "stop_id")
                ) ## filter stop_times to the key stops in each segment
    max_stop_times = max_stop_times.dropna(subset=['departure_time'])
    max_stop_times = max_stop_times.drop_duplicates(subset=['trip_id','hqta_segment_id']) ## filter duplicates for top2 approach
    
    if shape_id in debug_ids:
        debug_dict[f'{shape_id}_max_stop0'] = max_stop_times
    
    max_stop_times['departure_time'] = max_stop_times['departure_time'].apply(fix_arrival_time) ## reformat GTFS time to a format datetime can ingest
    max_stop_times['departure_dt'] = max_stop_times['departure_time'].apply(lambda x:
                                                                    dt.datetime.strptime(x, '%H:%M:%S'))
    max_stop_times['departure_hour'] = max_stop_times['departure_dt'].apply(lambda x: x.hour)
    
    # max_stop_times = max_stop_times >> filter(_.departure_hour.isin(commute_hours))
    
    if max_stop_times.size == 0:
        print(f'no commute hour trips for shape {shape_id}')
        return
    
    if shape_id in debug_ids:
        debug_dict[f'{shape_id}_max_stop'] = max_stop_times

    ## new flexible peak
    segment_am_max = (max_stop_times
                  >> count(_.hqta_segment_id, _.departure_hour)
                  >> filter(_.departure_hour < 12)
                  >> group_by(_.hqta_segment_id)
                  >> summarize(am_max_trips = _.n.max())
                 )
    
    segment_pm_max = (max_stop_times
                  >> count(_.hqta_segment_id, _.departure_hour)
                  >> filter(_.departure_hour >= 12)
                  >> group_by(_.hqta_segment_id)
                  >> summarize(pm_max_trips = _.n.max())
                 )
    try:
        segment_peak_service = (segment_am_max >> inner_join(_, segment_pm_max, on = 'hqta_segment_id'))
        segment_peak_service['hq_transit_corr'] = segment_peak_service.apply(
                                    lambda x: x.am_max_trips > 4 and x.pm_max_trips > 4, axis = 1)
    except: ## append when all segments only have am or pm trips, not an hqta by definition
        segment_peak_service = segment_am_max.append(segment_pm_max)
        segment_peak_service['hq_transit_corr'] = False
    ##
    
    segment_peak_service = segment_peak_service.reset_index(drop=True)
    segment_peak_service['last_seg_hqta'] = segment_peak_service['hq_transit_corr'].shift(1)
    segment_peak_service['next_seg_hqta'] = segment_peak_service['hq_transit_corr'].shift(-1)
    
    ## consider segment not HQTA if both the prior and next segements are not HQTAs
    segment_peak_service['hq_transit_corr'] = segment_peak_service.apply(
            lambda x: False if x.hq_transit_corr == False else x.last_seg_hqta or x.next_seg_hqta, axis = 1)
    segment_peak_service['hq_transit_corr'] = segment_peak_service['hq_transit_corr'].fillna(True)
    
    single_hqta = segments_with_max_stop >> inner_join(_, segment_peak_service, on = 'hqta_segment_id')
    single_hqta.drop(columns=['calitp_extracted_at', 'next_seg_hqta', 'last_seg_hqta',
                               'index', 'n_trips'], inplace=True)
    
    if shape_id in debug_ids:
        debug_dict[f'{shape_id}_single_hqta'] = single_hqta
    
    return single_hqta.reset_index(drop=True)

In [31]:
def single_operator_hqta(views):  
    
    global debug_dict
    debug_dict = {}
    
    shapes, trips, stop_times, stops = views
    
    
    distinct_routes = (trips
                   >> distinct(_.route_id, _.shape_id, _.direction_id, _keep_all = True)
                   >> select(_.calitp_itp_id, _.route_id, _.shape_id, _.direction_id, _.trip_id))
    
    route_count_by_stop = (stop_times
                         >> select(_.stop_id, _.trip_id)
                         >> inner_join(_, distinct_routes, on = "trip_id")
                         >> count(_.stop_id)
                         >> rename(n_routes = _.n)
                         >> arrange(-_.n_routes))    
    
    hqta = gpd.GeoDataFrame()
    ## start with shapes including the highest number of trips
    shapes_sorted = trips.groupby('shape_id').count().sort_values(by='trip_id', ascending=False).index
    shapes_sorted = pd.Series(shapes_sorted)
    total_shapes = len(shapes_sorted)
    print(f'there are {total_shapes} shapes total')
    for ix, shape_id in shapes_sorted.items():
        print(f'calculating for shape_id {shape_id}')
        if ix % 25 == 0:
            print(f'progress: {ix}/{total_shapes} shapes ({round(((ix/total_shapes)*100), 2)}%)')
        try:
            result = single_shape_hqta(shapes, trips, stop_times, stops, route_count_by_stop,
                                    shape_id, hqta)
            hqta = hqta.append(result)
        except:
            print(f'unable to calculate HQTA for shape_id {shape_id}')
        try:
            hqta = hqta.set_crs('EPSG:6414')
        except:
            continue
    return hqta
    # return hqta.drop(columns=['n', 'departure_hour']).reset_index(drop=True)

In [27]:
debug_ids = ['940143']

In [21]:
lbt_views = get_operator_views(170)

loaded shapes
loaded bus routes
loaded trips
loaded stop times
loaded stops


In [32]:
lbt_hqta = single_operator_hqta(lbt_views)

there are 100 shapes total
calculating for shape_id 510066
progress: 0/100 shapes (0.0%)
calculating for shape_id 610106
calculating for shape_id 610105
calculating for shape_id 510067
calculating for shape_id 1210143
calculating for shape_id 460064
calculating for shape_id 1210138
calculating for shape_id 460067
no line for shape 460067
calculating for shape_id 1920161
calculating for shape_id 1920167
calculating for shape_id 1730123
calculating for shape_id 1910292
calculating for shape_id 1910295
calculating for shape_id 1730122
calculating for shape_id 220068
calculating for shape_id 220074
calculating for shape_id 1710168
calculating for shape_id 1710162
calculating for shape_id 1720090
calculating for shape_id 1720091
no line for shape 1720091
calculating for shape_id 10070
calculating for shape_id 20008
calculating for shape_id 40001
calculating for shape_id 40002
no line for shape 40002
calculating for shape_id 410006
calculating for shape_id 80002
progress: 25/100 shapes (25.0

In [33]:
lbt_hqta

calitp_itp_id                                           geometry  \
0            170  POLYGON ((167654.764 -470242.512, 167655.638 -...   
1            170  POLYGON ((167694.221 -469153.485, 167693.191 -...   
2            170  POLYGON ((167669.771 -467897.042, 167669.770 -...   
3            170  POLYGON ((167646.521 -466641.120, 167646.502 -...   
4            170  POLYGON ((167605.970 -464592.739, 167604.537 -...   
..           ...                                                ...   
8            170  POLYGON ((167565.314 -466042.457, 167566.287 -...   
9            170  POLYGON ((167563.316 -466013.754, 167563.368 -...   
10           170  POLYGON ((167564.027 -466023.335, 167507.511 -...   
11           170  POLYGON ((167584.688 -466111.742, 167498.932 -...   
0            170  POLYGON ((165145.529 -464024.944, 165145.017 -...   

    hqta_segment_id segment_sequence shape_id stop_id  am_max_trips  \
0        1081235767                0   510066    4010            16   
1         930441633                1   510066    0004             7   
2        2927376411                2   510066    1698             7   
3        3648718989                3   510066    0550             4   
4        1193254190                4   510066    0555             4   
..              ...              ...      ...     ...           ...   
8        2433783489               42  4050003    5251             1   
9        3860309591               43  4050003    5251             1   
10        259261282               45  4050003    0610             5   
11       2524795608               46  4050003    5251             1   
0        1235633368                4  4050001    0134             2   

    pm_max_trips  hq_transit_corr  
0             16             True  
1              7             True  
2              7             True  
3              4            False  
4              4            False  
..           ...              ...  
8              1            False  
9              1            False  
10             4            False  
11             1            False  
0              2            False  

[383 rows x 9 columns]

In [36]:
lbt_hqta >> filter(_.stop_id == '0477')

calitp_itp_id                                           geometry  \
3           170  POLYGON ((170683.457 -468570.782, 170703.916 -...   

   hqta_segment_id segment_sequence shape_id stop_id  am_max_trips  \
3       3462120608                4   460064    0477             6   

   pm_max_trips  hq_transit_corr  
3             6             True

In [41]:
lbt_hqta.info()

<class 'geopandas.geodataframe.GeoDataFrame'>
Int64Index: 383 entries, 0 to 0
Data columns (total 9 columns):
 #   Column            Non-Null Count  Dtype   
---  ------            --------------  -----   
 0   calitp_itp_id     383 non-null    object  
 1   geometry          383 non-null    geometry
 2   hqta_segment_id   383 non-null    int64   
 3   segment_sequence  383 non-null    object  
 4   shape_id          383 non-null    object  
 5   stop_id           383 non-null    object  
 6   am_max_trips      383 non-null    int64   
 7   pm_max_trips      383 non-null    int64   
 8   hq_transit_corr   383 non-null    bool    
dtypes: bool(1), geometry(1), int64(3), object(4)
memory usage: 27.3+ KB


In [43]:
lbt_hqta.shape_id.unique()

array(['510066', '610106', '610105', '510067', '1210143', '460064',
       '1210138', '1920161', '1920167', '1730123', '1910292', '1910295',
       '1730122', '220068', '220074', '1710168', '1710162', '1720090',
       '10070', '20008', '40001', '410006', '80002', '1510067', '20010',
       '710048', '1510064', '450110', '710046', '1040017', '1310131',
       '1040018', '910129', '1010081', '1030063', '910132', '1810080',
       '1030064', '230018', '230015', '1120045', '1110083', '210080',
       '1820070', '1810081', '920113', '930151', '4050003', '4050001'],
      dtype=object)

In [40]:
map_hqta(lbt_hqta, 'hqta_segment_id')

Map(center=[33.77270983724983, -118.1893899136652], controls=(ZoomControl(options=['position', 'zoom_in_text',…

In [44]:
map_hqta(lbt_hqta >> filter(_.shape_id == '1730123'))

Map(center=[33.91231602534506, -118.10067122142928], controls=(ZoomControl(options=['position', 'zoom_in_text'…

In [ ]:
## investigate PCH not being an hqta...

In [45]:
lbt_hqta >> filter(_.shape_id == '1730123')

calitp_itp_id                                           geometry  \
0            170  POLYGON ((175350.451 -454051.814, 175349.213 -...   
1            170  POLYGON ((175836.936 -454587.721, 175836.959 -...   
2            170  POLYGON ((175798.075 -455972.107, 175798.083 -...   
3            170  POLYGON ((175827.602 -457076.012, 175827.610 -...   
4            170  POLYGON ((175854.103 -458921.667, 175854.135 -...   
5            170  POLYGON ((176737.013 -460049.998, 176737.342 -...   
6            170  POLYGON ((177991.832 -460007.766, 178010.111 -...   
7            170  POLYGON ((178396.327 -461576.519, 178395.927 -...   
8            170  POLYGON ((178429.971 -462989.509, 178430.591 -...   
9            170  POLYGON ((178152.046 -463095.761, 178090.699 -...   
10           170  POLYGON ((175975.412 -463280.594, 175977.308 -...   
11           170  POLYGON ((175919.561 -463405.079, 175916.534 -...   
12           170  POLYGON ((176053.775 -464584.739, 176053.777 -...   
13           170  POLYGON ((176068.527 -465823.921, 176068.537 -...   
14           170  POLYGON ((175997.767 -467784.889, 175997.533 -...   
15           170  POLYGON ((175564.080 -467890.219, 175563.606 -...   
16           170  POLYGON ((173784.416 -467841.455, 173785.048 -...   
17           170  POLYGON ((171789.381 -467743.362, 171793.498 -...   
18           170  POLYGON ((171187.739 -467856.064, 171187.546 -...   
19           170  POLYGON ((169935.539 -467879.393, 169923.522 -...   
20           170  POLYGON ((168696.051 -467903.956, 168680.399 -...   
21           170  POLYGON ((168037.524 -467916.250, 168037.372 -...   
22           170  POLYGON ((167357.829 -467928.625, 167358.227 -...   
23           170  POLYGON ((167281.629 -469235.162, 167281.182 -...   

    hqta_segment_id segment_sequence shape_id stop_id  am_max_trips  \
0        2631631551                0  1730123    4113             5   
1        3957109289                1  1730123    4116             3   
2        1926620051                2  1730123    4119             3   
3          97702661                3  1730123    3033             3   
4        2612414118                4  1730123    4124             3   
5        3971053104                5  1730123    1909             3   
6        1975003018                6  1730123    1902             2   
7          46077724                7  1730123    1569             3   
8        2449476237                8  1730123    1802             5   
9        3842456091                9  1730123    1272             4   
10       3062012927               10  1730123    1972             4   
11       3246762857               11  1730123    1795             2   
12       1485634259               12  1730123    1792             2   
13        797690437               13  1730123    1788             2   
14       2985254886               14  1730123    0529             2   
15       3337121648               15  1730123    0527             2   
16       1608597194               16  1730123    1779             2   
17        686165596               17  1730123    1346             7   
18       3092845517               18  1730123    1672             7   
19       3479053147               19  1730123    1667             7   
20       2645542972               20  1730123    1663             7   
21       3936942250               21  1730123    1662             7   
22       1939974416               22  1730123    0054             7   
23         78027142               23  1730123    0360            10   

    pm_max_trips  hq_transit_corr  
0              5             True  
1              2            False  
2              2            False  
3              2            False  
4              2            False  
5              3            False  
6              2            False  
7              3            False  
8              5            False  
9              3            False  
10             3            False  
11             2   

## Multiple Operators

In [11]:
itp_ids = tbl.gtfs_schedule.agency() >> distinct(_.calitp_itp_id) >> collect()
itp_ids = itp_ids.calitp_itp_id

In [12]:
def multiple_operator_hqta(existing=pd.DataFrame()):
    
    hqta = pd.DataFrame()    
    for _, itp_id in itp_ids.items():
        if not existing.empty:
            if str(itp_id) in existing['calitp_itp_id'].to_list():
                print(f'{itp_id} existing')
                # break
                continue
        if int(itp_id) == 200:
            continue ## skip MTC feed to use individual operator feeds
        try:
            print(f'attempting for operator {itp_id}')
            hqta = hqta.append(single_operator_hqta(get_operator_views(itp_id)))
        except:
            print(f'failed for operator {itp_id}')
        
    return hqta

In [6]:
# all_operators = multiple_operator_hqta()
# geoparquet_gcs_export(all_operators, 'all_operators')

In [3]:
all_operators = gpd.read_parquet(f'{GCS_FILE_PATH}all_operators.parquet')

In [14]:
itp_ids = itp_ids.astype(str)

In [15]:
df = tbl.gtfs_schedule.agency() >> collect() >> distinct(_.calitp_itp_id, _keep_all = True)

In [16]:
not_ran = df[~df['calitp_itp_id'].isin(all_operators['calitp_itp_id'].astype('int64'))]

In [17]:
with pd.option_context('display.max_rows', 100):
    display(not_ran)

calitp_itp_id  calitp_url_number                             agency_id  \
2              474                  0                                 TCRTA   
3              256                  0                           Porterville   
4              279                  0                                  BART   
5              200                  0                                    BA   
24             325                  0                                    68   
25              62                  0                                    68   
27             372                  0  d139e461-175c-4156-beb0-e6a23b64d978   
36              13                  0                                  1206   
38             338                  0                                   513   
42              10                  0                                    CE   
43             246                  0                                  1000   
86             254                  0                                    20   
90              15                  0                                   897   
91              41                  0                                   897   
94             323                  0                             Metrolink   
96             349                  0                                  1642   
97             315                  0                                  1535   
106            341                  0                                  1490   
116            280                  0                                   873   
117            271                  0                             Roseville   
137            344                  0                                    34   
143              7                  0                                   509   
153            312                  0                              11214031   
165            111                  0                                   FSL   
166            265                  0                                   989   
168            121                  0                                  1664   
169             97                  0                                   484   
175            122                  0                                  1707   

                               agency_name  \
2    Tulare County Regional Transit Agency   
3                      Porterville Transit   
4                   Bay Area Rapid Transit   
5                   Bay Area Rapid Transit   
24             Stanislaus Regional Transit   
25             Stanislaus Regional Transit   
27     Yolo County Transportation District   
36                                    None   
38                     Tideline Water Taxi   
42               Altamont Corridor Express   
43                                Caltrain   
86                          Plumas Transit   
90              Angel Island Tiburon Ferry   
91              Angel Island Tiburon Ferry   
94                        Metrolink Trains   
96                         Turlock Transit   
97          Sonoma Marin Area Rail Transit   
106                                 TRACER   
116                San Francisco Bay Ferry   
117                      Roseville Transit   
137                        Trinity Transit   
143                             Spirit Bus   
153                              A11214029   
165                     Folsom Stage Lines   
166                        Blossom Express   
168       Glendora Transportation Division   
169                         Duarte Transit   
175                  Glenn Transit Service   

                                            agency_url      agency_timezone  \
2                                 https://ridetcat.org  America/Los_Angeles   
3                                http://www.ridept.com  America/Los_Angeles   
4                                https://www.bart.gov/  America/Los_Angeles   
5                                https://www.bart.gov/  Americ

## Spatially Aggregate non-HQTAs

* currently unable with new flexible peak definition

In [26]:
# non_hqta = all_operators >> filter(-_.hq_transit_corr)

# non_hqta_overlaid = non_hqta.overlay(non_hqta, how='intersection')

# non_hqta_deduplicated = non_hqta_overlaid[non_hqta_overlaid['hqta_segment_id_1'] != non_hqta_overlaid['hqta_segment_id_2']]

In [27]:
# def combine_segment_ids(row):
#     both_segments = [row.hqta_segment_id_1, row.hqta_segment_id_2]
#     both_segments.sort()
#     return str(both_segments)

In [28]:
# non_hqta_deduplicated['overlap_id'] = non_hqta_deduplicated.apply(combine_segment_ids, axis=1)

In [44]:
# more_deduplicated = non_hqta_deduplicated.drop_duplicates(subset=['overlap_id'])

In [8]:
# more_deduplicated.to_parquet('./data/bus/overlaps.parquet')

In [47]:
# ##TODO rewrite to new definition... (or just drop?)

# def sum_overlaps(row):
#     row['am_peak'] = row.am_peak_1 + row.am_peak_2
#     row['pm_peak'] = row.pm_peak_1 + row.pm_peak_2
#     row['n_trips'] = row.n_trips_1 + row.n_trips_2
#     if row.am_peak >= 4 and row.pm_peak >= 4:
#         row['hq_transit_corr'] = True
#     else:
#         row['hq_transit_corr'] = False
#     return row

In [48]:
# overlaps_summed = more_deduplicated.apply(sum_overlaps, axis = 1) 

In [49]:
# new_hq = overlaps_summed[overlaps_summed['hq_transit_corr']]

In [50]:
# new_hq = new_hq[new_hq['am_peak'] != new_hq['am_peak_1']*2] ## drop likely duplicates; can rework above to do this better

In [51]:
# new_hq = new_hq.drop(columns=['calitp_itp_id_2',
#                'hqta_segment_id_2', 'index_2', 'n_trips_2', 'segment_sequence_2',
#                'shape_id_2', 'stop_id_2', 'am_peak_2', 'pm_peak_2',
#                'hq_transit_corr_2', 'index_1', 'index_2', 'n_trips_2', 'n_trips_1',
#                 'am_peak_1', 'pm_peak_1', 'hq_transit_corr_1'])

In [52]:
# new_hq = new_hq.rename(columns = {'calitp_itp_id_1':'calitp_itp_id', 'hqta_segment_id_1':'hqta_segment_id',
#        'segment_sequence_1':'segment_sequence', 'shape_id_1':'shape_id', 'stop_id_1':'stop_id'})

In [55]:
# all_operators = all_operators.append(new_hq)

In [4]:
short_dropped = all_operators[all_operators['geometry'].area > 50*400] ##50m width * 400m segment min

In [5]:
short_dropped = short_dropped.reset_index().drop(columns=['index'])

In [6]:
# short_dropped.to_parquet('./data/bus/all_operators_cleaned_appended.parquet')

In [7]:
dissolved = short_dropped.dissolve(by=['hq_transit_corr', 'shape_id']).reset_index()

In [8]:
dissolved = dissolved[dissolved['geometry'].area > 50*3000] ##50m width * 3000m shape min

In [37]:
dissolved = gpd.read_parquet(f'{GCS_FILE_PATH}shape_hqta_dissolve.parquet')

In [38]:
map_hqta(dissolved)

Map(center=[33.596015707487744, -117.87571863402819], controls=(ZoomControl(options=['position', 'zoom_in_text…

In [14]:
geoparquet_gcs_export(dissolved, 'shape_hqta_dissolve')

/home/jovyan/data-analyses/high_quality_transit_areas/utilities.py:85: UserWarning: this is an initial implementation of Parquet/Feather file support and associated metadata.  This is tracking version 0.1.0 of the metadata specification at https://github.com/geopandas/geo-arrow-spec

This metadata specification does not yet make stability promises.  We do not yet recommend using this in a production setting unless you are able to rewrite your Parquet/Feather files.

To further ignore this warning, you can do: 
import warnings; warnings.filterwarnings('ignore', message='.*initial implementation of Parquet.*')
  gdf.to_parquet(f"{name}.parquet")


### Bay Area Map Image

![map](img/bay.png)

### Los Angeles Map Image

![map](img/la.png)

## Result

* Ran and aggregated for nearly all bus operators statewide
    * Segments not containing stops will not appear as HQTA-- may need to interpolate (e.g., freeway segments)
    * Some questionable short segments
* Algorithm may be overestimating for SFMTA

### Data Issues

* 61 County Connection has a lot of na departure times?
    * strange results after dropna
* 48 B-Line similarly choppy
* 116 Fresno Area Express has whitespace in departure times
    * ValueError: time data ' 7:04:00' does not match format '%H:%M:%S'

